### PR TITLE
Fix: config API - autoresume & resume

### PR DESCRIPTION
**Pull request by ditwoo**
## Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)?
- [x] Did you check the code style? `catalyst-make-codestyle && catalyst-check-codestyle` (`pip install -U catalyst-codestyle`).
- [ ] Did you make sure to update the docs? We use Google format for all the methods and classes.
- [ ] Did you check the docs with `make check-docs`?
- [x] Did you write any new necessary tests?
- [ ] Did you add your new functionality to the docs?
- [x] Did you update the [CHANGELOG](https://github.com/catalyst-team/catalyst/CHANGELOG.md)?
- [x] **You can use 'Login as guest' to see Teamcity build logs.**

## Description

Fix related to this issue [904](https://github.com/catalyst-team/catalyst/issues/904).

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: tests/_tests_dl_callbacks/experiment.py [hunk 1]
```diff
@@ -1,13 +1,30 @@
...
 from torch.utils.data import Subset
 
+from catalyst import dl, registry
 from catalyst.contrib.datasets import MNIST
 from catalyst.data.cv import Compose, Normalize, ToTensor
 from catalyst.dl import ConfigExperiment
 
 
+@registry.Callback
+class KeyboardInteruptCallback(dl.Callback):
```
Comment (Scitator) [inline tests/_tests_dl_callbacks/experiment.py]: let's move it to contrib 😂

File: tests/_tests_dl_callbacks/config21.yml [hunk 2]
```diff
@@ -0,0 +1,50 @@
+model_params:
```
Comment (Scitator) [inline tests/_tests_dl_callbacks/config21.yml]: do we really need new config? could we just reuse some of the old ones?

File: catalyst/dl/experiment/config.py [hunk 3]
```diff
@@ -479,6 +478,15 @@ def _process_callbacks(callbacks: OrderedDict) -> None:
         """
         for callback in callbacks.values():
             if isinstance(callback, CheckpointCallback):
+                # use resume only for first stage
+                _resume = self._config["args"].pop("resume", None)
```
Comment (Scitator) [inline catalyst/dl/experiment/config.py]: let's move it here
https://github.com/catalyst-team/catalyst/blob/master/catalyst/core/callbacks/checkpoint.py#L553
what do you think about it?

File: bin/tests/check_dl_core_callbacks.sh [hunk 4]
```diff
@@ -789,4 +789,36 @@ assert metrics['last']['cmc01'] > 0.1  # slightly better then random
...
+LOG_MSG='pipeline 22'
+echo ${LOG_MSG}
+
+EXPDIR=./tests/_tests_dl_callbacks
+LOGDIR=./tests/logs/_tests_dl_callbacks
+CHECKPOINTS=${LOGDIR}/checkpoints
+LOGFILE=${CHECKPOINTS}/_metrics.json
+EXP_OUTPUT=./tests/output.txt
+
+INTERRUPT_EPOCH=4 \
```
Comment (Scitator) [inline bin/tests/check_dl_core_callbacks.sh]: could we use `--check` arg instead?

File: catalyst/dl/experiment/config.py (hunk 3 above)
Comment (ditwoo) [reply catalyst/dl/experiment/config.py]: Current implementation updates arguments of CheckpointCallback.
Could you please describe more about what's exactly need to move to callback?

File: tests/_tests_dl_callbacks/config21.yml (hunk 2 above)
Comment (ditwoo) [reply tests/_tests_dl_callbacks/config21.yml]: thanks,
nice point, probably we can update one of previous config files - add KeyboardInterrupt callback and use env variable for interrupting.

File: bin/tests/check_dl_core_callbacks.sh (hunk 4 above)
Comment (ditwoo) [reply bin/tests/check_dl_core_callbacks.sh]: hmm, config already have CheckRunCallback, i think it's a same thing

File: tests/_tests_dl_callbacks/experiment.py (hunk 1 above)
Comment (ditwoo) [reply tests/_tests_dl_callbacks/experiment.py]: Are you sure?)
This callback is only for tests)

File: catalyst/dl/experiment/config.py (hunk 3 above)
Comment (Scitator) [reply catalyst/dl/experiment/config.py]: I mean, based on [this code](github.com/catalyst-team/catalyst/blob/master/catalyst/core/callbacks/checkpoint.py#L548#L558) `CheckpointCallback` already uses arguments from state/runner like `resume` and `resume-dir`
so, could we just propagate them there? what do you think about it?

File: tests/_tests_dl_callbacks/experiment.py [hunk 5]
```diff
@@ -1,5 +1,6 @@
 # flake8: noqa
 from collections import OrderedDict
+import os
```
Comment (Scitator) [inline tests/_tests_dl_callbacks/experiment.py]: do we need os here?

File: catalyst/dl/experiment/config.py [hunk 6]
```diff
@@ -468,17 +468,19 @@ def _get_callback(**params):
         return callback
 
     @staticmethod
-    def _process_callbacks(callbacks: OrderedDict) -> None:
+    def _process_callbacks(callbacks: OrderedDict, stage_num: int) -> None:
```
Comment (Scitator) [inline catalyst/dl/experiment/config.py]: `stage_index`?

File: catalyst/dl/experiment/config.py [hunk 7]
```diff
@@ -537,7 +539,8 @@ def get_callbacks(self, stage: str) -> "OrderedDict[Callback]":
             if not is_already_present:
                 callbacks[callback_name] = callback_fn()
 
-        self._process_callbacks(callbacks)
+        stage_number = list(self.stages_config.keys()).index(stage)
```
Comment (Scitator) [inline catalyst/dl/experiment/config.py]: because you can rename it to `stage_index` too

Comment (Scitator) [reply catalyst/dl/experiment/config.py]: and it could be a bit easier to understand

File: catalyst/dl/experiment/config.py (hunk 6 above)
Comment (Scitator) [reply catalyst/dl/experiment/config.py]: I also think it should be `None` by default

File: catalyst/dl/experiment/config.py [hunk 8]
```diff
@@ -468,17 +468,19 @@ def _get_callback(**params):
...
         run of config experiment.
 
         Arguments:
             callbacks (OrderedDict): finalized order of callbacks.
+            stage_num (int): number of a current stage
         """
+
         for callback in callbacks.values():
-            if isinstance(callback, CheckpointCallback):
+            if isinstance(callback, CheckpointCallback) and stage_num > 0:
```
Comment (Scitator) [inline catalyst/dl/experiment/config.py]: could you please also add a comment here what is this hack for?

File: catalyst/core/callbacks/checkpoint.py [hunk 9]
```diff
@@ -545,6 +545,14 @@ def on_stage_start(self, runner: IRunner) -> None:
             from torch import save
         self._save_fn = save
 
+        if hasattr(runner, "resume") and runner.resume:
```
Comment (Scitator) [inline catalyst/core/callbacks/checkpoint.py]: `getattr(runner, "resume",  None) is not None`?

File: catalyst/core/callbacks/checkpoint.py [hunk 10]
```diff
@@ -545,6 +545,14 @@ def on_stage_start(self, runner: IRunner) -> None:
...
+            self.resume = runner.resume
+            runner.resume = None
+        elif hasattr(runner, "autoresume") and runner.autoresume:
```
Comment (Scitator) [inline catalyst/core/callbacks/checkpoint.py]: the same here :)

**Pull request merged**